pkg/cluster/mutex.go

```diff
@@ -31,20 +32,35 @@ type Mutex interface {
 }
 
 type mutex struct {
+	// concurrency.Mutex is a session level mutex, so sync.Mutex is
+	// required to make it goroutine safe
+	lock    sync.Mutex
 	m       *concurrency.Mutex
 	timeout time.Duration
 }
 
-func (m *mutex) Lock() error {
+func (m *mutex) Lock() (err error) {
+	panicked := true
+
+	m.lock.Lock()
+	defer func() {
+		if panicked || err != nil {
```

Comment:
(nit) Using `recover` is another option, this way is fine to me.

Reply:
This function is unable to handle the panic completely, so using `recover` requires `panic` again to expose the panic to callers, but this will change the original stack trace.

Reply:
this looks cool